### PR TITLE
Allow adding more CPUs than the max sockets count

### PR DESCRIPTION
**Pull request by bond95**
Added division of CPU number in case if it bigger then max allowed number of sockets.

Fixes: https://github.com/oVirt/ovirt-web-ui/issues/924

### CODE REVIEW

File: src/components/VmDetails/cards/DetailsCard/index.js [hunk 1]
```diff
@@ -421,13 +422,30 @@ class DetailsCard extends React.Component {
           break
 
         case 'cpu':
-          if (isNumber(value) && value > 0) {
+          if (isNumber(value) && value > 0 && value < this.props.config.get('maxNumOfVmCpus')) {
```
Comment (michalskrivanek) [inline src/components/VmDetails/cards/DetailsCard/index.js]: <=?

Comment (bond95) [reply src/components/VmDetails/cards/DetailsCard/index.js]: Oh, yeah, right

File: src/components/VmDetails/cards/DetailsCard/index.js [hunk 2]
```diff
@@ -421,13 +422,30 @@ class DetailsCard extends React.Component {
...
+            }
             updates = updates.mergeDeep({
               'cpu': {
                 'topology': { // NOTE: this may not match VM's template's topology
-                  'sockets': +value,
-                  'cores': 1,
-                  'threads': 1,
+                  sockets,
+                  cores,
+                  threads,
```
[the remark this answers is not in the document]
Comment (gregsheremeta) [reply src/components/VmDetails/cards/DetailsCard/index.js]: selection_016

i think x86 are 1 or 2, but power can go up to 8

**Comment by sjd78** [comment]
@bond95 - 
For `src/utils/index.js`:
````js
export function balanceCpuTopology (totalCPUs: number, maxSockets: number, maxCPUs: number): 
  sockets: number,
  cores: number,
  threads: number
} {
  if (totalCPUs > maxCPUs) {
    throw new Error(`totalCPUs ${totalCPUs} > maxCPUs ${maxCPUs}`)
  }

  let sockets = totalCPUs
  let cores = 1
  let threads = 1

  if (totalCPUs > maxSockets) {
    const largestFactorUnderMax = dividers(totalCPUs).reduce((prev, curr) => curr > maxSockets ? rev : Math.max(prev, curr))

    sockets = largestFactorUnderMax
    cores = dividers(totalCPUs / sockets).pop()
  }

  return {
    sockets,
    cores,
    threads,
  }
}

````

For `src/utils/index.test.js`:
````js
/* eslint-env jest */

import { dividers, balanceCpuTopology } from './index'

describe('test generic ', () => {
  it('test dividers', () => {
    const sampleData = [ 1, 5, 10, 14, 15, 19, 22, 78, 99, 124 ]
    const sampleResults = [
      [ 1 ],
      [ 1, 5 ],
      [ 1, 2, 5, 10 ],
      [ 1, 2, 7, 14 ],
      [ 1, 3, 5, 15 ],
      [ 1, 19 ],
      [ 1, 2, 11, 22 ],
      [ 1, 2, 3, 6, 13, 26, 39, 78 ],
      [ 1, 3, 9, 11, 33, 99 ],
      [ 1, 2, 4, 31, 62, 124 ],
    ]

    for (let i = 0; i < sampleData.length; i++) {
      const divs = dividers(sampleData[i])
      expect(divs).toEqual(sampleResults[i])
    }
  })

  it('test balance CPU topology', () => {
    const maxSockets = 16
    const maxCPUs = 384
    const data = [
      { ask: 1, r: { sockets: 1, cores: 1, threads: 1 } },
      { ask: 7, r: { sockets: 7, cores: 1, threads: 1 } },
      { ask: 18, r: { sockets: 9, cores: 2, threads: 1 } },
      { ask: 19, r: { sockets: 1, cores: 19, threads: 1 } },
      { ask: 29, r: { sockets: 1, cores: 29, threads: 1 } },
      { ask: 42, r: { sockets: 14, cores: 3, threads: 1 } },
      { ask: 96, r: { sockets: 16, cores: 6, threads: 1 } },
      { ask: 100, r: { sockets: 10, cores: 10, threads: 1 } },
      { ask: 234, r: { sockets: 13, cores: 18, threads: 1 } },
    ]

    for (let entry of data) {
      const balanced = balanceCpuTopology(entry.ask, maxSockets, maxCPUs)
      expect(balanced).toEqual(entry.r)
    }
  })
})
````

File: src/components/VmDetails/cards/DetailsCard/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjd78) [reply src/components/VmDetails/cards/DetailsCard/index.js]: > i think x86 are 1 or 2, but power can go up to 8

Yeah, so I figure for web-ui, the threads can always be `1` then.

[the remark this answers is not in the document]
Comment (gregsheremeta) [reply src/components/VmDetails/cards/DetailsCard/index.js]: I don't follow. We could have a power host and template right? Don't we want to read the values from the template? I might be missing something 🤔

[the remark this answers is not in the document]
Comment (michalskrivanek) [reply src/components/VmDetails/cards/DetailsCard/index.js]: @sjd78 @gregsheremeta 
It should definitely honor the template. OTOH it's a fair consideration if we want to show threads at all or how. While they are useful on POWER8, on x86 it's almost entirely useless to use anything but 1
Maybe for the new design?
Here I don't mind keeping it as is, as long as the info icon explains to users that they shouldn't touch it

[the remark this answers is not in the document]
Comment (gregsheremeta) [reply src/components/VmDetails/cards/DetailsCard/index.js]: ( these max values are in vdc_options - MaxNumOfCpuPerSocket, MaxNumOfThreadsPerCpu, MaxNumOfVmCpus, MaxNumOfVmSockets )

[the remark this answers is not in the document]
Comment (sjd78) [reply src/components/VmDetails/cards/DetailsCard/index.js]: The screenshot Greg posted is from webadmin.  On web-ui, we have 1 field called "CPUs".  At this point, we only show, and allow editing of, the aggregate CPU count.  Adding the "Advanced Parameters" to web-ui to have the users be able to tweak the sockets, cores and threads doesn't make much sense to me.

So in the case of a VM being based on a template (but not cloned from a template), I'm not quite sure how we honor the template's thread count when changing the CPU count.  Looks like webadmin doesn't try to retain the thread count.

[the remark this answers is not in the document]
Comment (sjd78) [reply src/components/VmDetails/cards/DetailsCard/index.js]: > MaxNumOfCpuPerSocket, MaxNumOfThreadsPerCpu, MaxNumOfVmCpus, MaxNumOfVmSocket

Via `/api/options/${option}`...
  - MaxNumOfVmCpus = 384
  - MaxNumOfVmSockets = 16
  - MaxNumOfCpuPerSocket =  254
  - MaxNumOfThreadsPerCpu = 8

So any prime number >254 and <384 isn't a valid CPU count.  (Did I do my number theory right?)

[the remark this answers is not in the document]
Comment (gregsheremeta) [reply src/components/VmDetails/cards/DetailsCard/index.js]: > The screenshot Greg posted is from webadmin. On web-ui, we have 1 field called "CPUs". At this point, we only show, and allow editing of, the aggregate CPU count

I confirmed with Michal that we should change to match admin portal. So we should allow setting threads, and if it's x86, we should default threads to 1 and also provide notice that using more than 1 on x86 isn't usually a good idea.

So we need to add the fields and some more complex logic -- logic that mimics how admin portal works.

**Comment by lwrigh** [comment]
@bond95 Is any design work needed for it?

**Comment by lwrigh** [comment]
Screen Shot 2019-03-14 at 11 29 58 AM
Screen Shot 2019-03-14 at 11 30 28 AM
@gregsheremeta @sjd78 @bond95 @michalskrivanek What do you guys think of this option for adding more CPU options to the details card?

**Comment by gregsheremeta** [comment]
> @gregsheremeta @sjd78 @bond95 @michalskrivanek What do you guys think of this option for adding more CPU options to the details card?

I like that it's nice and simple and matches admin portal. I do wonder about having two Advanced Settings right in the same card, but I'm not sure it would look right if we had the main CPU count up top and then the 3 sub fields in the bottom Advanced fold. @lwrigh would you mind attempting a render of that?

**Comment by lwrigh** [comment]
> > @gregsheremeta @sjd78 @bond95 @michalskrivanek What do you guys think of this option for adding more CPU options to the details card?
> 
> I like that it's nice and simple and matches admin portal. I do wonder about having two Advanced Settings right in the same card, but I'm not sure it would look right if we had the main CPU count up top and then the 3 sub fields in the bottom Advanced fold. @lwrigh would you mind attempting a render of that?

I can mock that option up for comparison. I think it might be a good option to have the advanced options opened at first so the user can see what's grouped under the advanced options and then collapse it if they would like to. It might also help to put an info tip next to the total virtual CPUs label that lets the user know what parts make up the total virtual CPUs.

**Comment by gregsheremeta** [comment]
> I think it might be a good option to have the advanced options opened at first so the user can see what's grouped under the advanced options and then collapse it if they would like to

The use case is valid, but then I'm nervous about how tall the card is :)

> It might also help to put an info tip next to the total virtual CPUs label that lets the user know what parts make up the total virtual CPUs.

great idea

**Comment by bond95** [comment]
@gregsheremeta @lwrigh @sjd78 

What do you think about moving topology fields to `Advanced options`?

image
image

**Comment by lwrigh** [comment]
@bond95 I tried to combine our suggestions into one design. What do you guys think? @sjd78 @gregsheremeta @bond95

**Comment by gregsheremeta** [comment]
> @bond95 I tried to combine our suggestions into one design. What do you guys think? @sjd78 @gregsheremeta @bond95 

Looks perfect. @michalskrivanek too - ?

**Comment by michalskrivanek** [comment]
in the second picture, how about just describe the current topology (since it's not next to it and by default hidden), something like 45 vCPUs - ok, so you have 9 sockets, 5 cores, 1 thread?

**Comment by lwrigh** [comment]
@michalskrivanek That's a good idea, so it would read something along the lines of 'Total virtual CPUs include virtual sockets, cores per virtual socket, and threads per core. You currently have 3 virtual sockets, 6 cores, and 1 thread.'

**Comment by michalskrivanek** [comment]
or a shorter "Total virtual CPUs include 3 virtual sockets, 6 cores per socket, and 1 thread per core"?:) 

Also the threads explanation - it should read x86_64, Power is spelled POWER, and since 4.2 we support P8 and P9 so the same generic architecture should be used, POWER: 8 (I think it doesn't make sense to use anything between 1-8 as the core can only run one virtual machine, so it makes sense to use them all otherwise they can't really be used)

**Comment by lwrigh** [comment]
@michalskrivanek That works for me if everyone is onboard with it. 

Okay so are you thinking of something along the lines of, "Please check your host architecture before changing the value. If you are not sure, it is recommended to keep threads per core set to 1. Other recommended values include x86_64: 1, POWER: 8." for the threads explanation?

**Comment by bond95** [comment]
> Also the threads explanation - it should read x86_64, Power is spelled POWER, and since 4.2 we support P8 and P9 so the same generic architecture should be used, POWER: 8 (I think it doesn't make sense to use anything between 1-8 as the core can only run one virtual machine, so it makes sense to use them all otherwise they can't really be used)

@michalskrivanek thread explanation was copied from Admin Portal, so I think it must changed there too?

**Comment by michalskrivanek** [comment]
@lwrigh "Please check your host architecture before changing the value. If you are not sure, it is recommended to keep threads per core set to 1 on x86_64 and 8 on POWER" ? or is it too short now?:)

**Comment by michalskrivanek** [comment]
> @michalskrivanek thread explanation was copied from Admin Portal, so I think it must changed there too

yes

**Comment by lwrigh** [comment]
> @lwrigh "Please check your host architecture before changing the value. If you are not sure, it is recommended to keep threads per core set to 1 on x86_64 and 8 on POWER" ? or is it too short now?:)

That wording works for me. I can update the wireframes to feature it.

**Comment by lwrigh** [comment]
Here are the updated wireframes: 
Screen Shot 2019-03-18 at 11 13 45 AM
Screen Shot 2019-03-18 at 11 13 56 AM
Screen Shot 2019-03-18 at 11 14 08 AM
Screen Shot 2019-03-18 at 11 14 20 AM
Screen Shot 2019-03-18 at 11 14 32 AM

**Comment by gregsheremeta** [comment]
@michalskrivanek Can we detect the host architecture via API? Then we could tailor the message and shrink it more.

**Comment by bond95** [comment]
> @michalskrivanek Can we detect the host architecture via API? Then we could tailor the message and shrink it more.

Yes, when VM portal get info about hosts, it also get info about it CPU: https://brq-dev.rhev.lab.eng.brq.redhat.com/ovirt-engine/apidoc/#/services/hosts/methods/list

**Comment by sjd78** [comment]
> > @michalskrivanek Can we detect the host architecture via API? Then we could tailor the message and shrink it more.
> 
> Yes, when VM portal get info about hosts, it also get info about it CPU: https://brq-dev.rhev.lab.eng.brq.redhat.com/ovirt-engine/apidoc/#/services/hosts/methods/list

Cluster is a better place to get the architecture as all of the hosts in a cluster need to be compatible (for VM migration).

**Comment by bond95** [comment]
> > > @michalskrivanek Can we detect the host architecture via API? Then we could tailor the message and shrink it more.
> > 
> > 
> > Yes, when VM portal get info about hosts, it also get info about it CPU: https://brq-dev.rhev.lab.eng.brq.redhat.com/ovirt-engine/apidoc/#/services/hosts/methods/list
> 
> Cluster is a better place to get the architecture as all of the hosts in a cluster need to be compatible (for VM migration).

Yes, but VM running on host, so

**Comment by bond95** [comment]
image

Tooltip for other then POWER architecture:
image

Tooltip for POWER architecture
image

@gregsheremeta @michalskrivanek @sjd78 @lwrigh

**Comment by lwrigh** [comment]
> image
> 
> Tooltip for other then POWER architecture:
> image
> 
> Tooltip for POWER architecture
> image
> 
> @gregsheremeta @michalskrivanek @sjd78 @lwrigh

@bond95 A couple comments on the wording, I would recommend changing the tooltip for other than POWER architecture to:
'Please check your host architecture to make sure it is properly set up before changing the value. If you are not sure, we recommend keeping threads-per-core set to 1. Other recommended values include: x86- 1, Power8- 1 through 8 (8 for high load VMs). '

I would also recommend changing the tooltip for POWER architecture to: 
'Your cluster architecture is POWER, we recommend using 1 through 8 values (8 for high load VMs).'

**Comment by bond95** [comment]
@lwrigh updated

image
image

**Comment by samuelmacko** [comment]
lgtm

**Pull request merged**